pkg/apis/configuration/v1beta1/kongupstreampolicy_types.go

```diff
@@ -0,0 +1,207 @@
+	// +kubebuilder:validation:Maximum=65536
+	Slots *int `json:"slots,omitempty"`
+
+	// HashOn defines how to calculate hash for consistent-hashing load balancing algorithm.
+	// Algorithm must be set to "consistent-hashing" for this field to have effect.
+	HashOn *KongUpstreamHash `json:"hash_on,omitempty"`
+
+	// HasOnFallback defines how to calculate hash for consistent-hashing load balancing algorithm if the primary hash
+	// function fails.
+	// Algorithm must be set to "consistent-hashing" for this field to have effect.
+	HashOnFallback *KongUpstreamHash `json:"hash_on_fallback,omitempty"`
+
+	// Healthchecks defines the health check configurations in Kong.
+	Healthchecks *KongUpstreamHealthcheck `json:"healthchecks,omitempty"`
+}
+
+// KongUpstreamHash defines how to calculate hash for consistent-hashing load balancing algorithm.
+// Only one of the fields must be set.
+type KongUpstreamHash struct {
+	// Header is the name of the header to use as hash input.
+	Header *string `json:"header,omitempty"`
+
+	// Cookie is the name of the cookie to use as hash input.
+	Cookie *string `json:"cookie,omitempty"`
+
+	// QueryArg is the name of the query argument to use as hash input.
+	QueryArg *string `json:"query_arg,omitempty"`
+
+	// URICapture is the name of the URI capture group to use as hash input.
+	URICapture *string `json:"uri_capture,omitempty"`
+}
+
+// KongUpstreamHealthcheck represents a health-check config of an Upstream in Kong.
+type KongUpstreamHealthcheck struct {
+	// Active configures active health check probing.
+	Active *KongUpstreamActiveHealthcheck `json:"active,omitempty"`
+
+	// Passive configures passive health check probing.
+	Passive *KongUpstreamPassiveHealthcheck `json:"passive,omitempty"`
+
+	// Threshold is the minimum percentage of the upstream’s targets’ weight that must be available for the whole
+	// upstream to be considered healthy.
+	Threshold *int `json:"threshold,omitempty"`
+}
+
+// KongUpstreamActiveHealthcheck configures active health check probing.
+type KongUpstreamActiveHealthcheck struct {
+	// Type determines how active health checks are collected.
+	Type *string `json:"type,omitempty"`
+
+	// Concurrency is the number of targets to check concurrently.
+	// +kubebuilder:validation:Minimum=1
+	Concurrency *int `json:"concurrency,omitempty"`
+
+	// Healthy configures thresholds and HTTP status codes to mark targets healthy for an upstream.
+	Healthy *KongUpstreamHealthcheckHealthy `json:"healthy,omitempty"`
+
+	// Unhealthy configures thresholds and HTTP status codes to mark targets unhealthy for an upstream.
+	Unhealthy *KongUpstreamHealthcheckUnhealthy `json:"unhealthy,omitempty"`
+
+	// HTTPPath is the path to use in GET HTTP request to run as a probe.
+	// +kubebuilder:validation:Pattern=^/.*$
+	HTTPPath *string `json:"http_path,omitempty"`
+
+	// HTTPSSni is the SNI to use in GET HTTPS request to run as a probe.
+	HTTPSSni *string `json:"https_sni,omitempty"`
```

Comment:
```suggestion
	HTTPSSNI *string `json:"https_sni,omitempty"`
```

With the caveat that it'll break generated stuff and probably should be done outside GitHub suggest.

This is unfortunately two acronyms stuck together, but I prefer having consistently capitalized acronyms in variables. `HTTPSServerName` or even just `ServerName` also work while looking a bit less silly. Breaking with the JSON name is fine.